src/cljs/owlet_ui/components/branch.cljs

```diff
@@ -4,6 +4,20 @@
             [reagent.core :as reagent :refer [atom]]
             [camel-snake-kebab.core :refer [->kebab-case]]))
 
+(def hover-image (reagent/atom ""))
+
+(defn get-random [images]
+  (get images (rand-int (count images))))
+
+(defn set-hover [images]
+  (let [image (get-random images)]
+    (if (> (count images) 1)
+      ; prevent the same image twice in a row
+      (if (== image @hover-image)
+        (set-hover images)
```

Comment:
Yup, JavaScript doesn't have TCO either. Do this in the browser console:

```
    >  function f() { f() }
    undefined
    >  f()
    VM99:1 Uncaught RangeError: Maximum call stack size exceeded
```

Re. 2.: I agree, in fact TCO was invented for Scheme. I'm not saying it's "wrong" to use ordinary recursion. In fact, you may often _have_ to do it if the function makes use of the return value. But I assert that recur, though created for Clojure out of practical necessity, turns out to be more readable, as it makes the program flow more explicit. And when combined with loop, it's more succinct as well, since you don't need to wrap the loop's contents in a named function.

I've actually found myself missing recur in Haskell, which _does_ have TCO. In fact, the post by Howard Abrams you referenced agrees!

> At first, I didn’t care much for the special consideration for dealing with tail recursion that just happens in other Lisp languages, however, I’ve changed my mind for the following reasons:
> 
> First, we often have to have an accumulator which we hide inside an inner function or lambda, the loop simply replaces what we normally do anyway.
> 
> Second, while you may want tail recursion, you may not be sure you are getting tail recursion. The Clojure approach ensures that recur is the last call in the function, and throws an error if not.
> 
> What appeared to me to be a kludge at first, has become a feature to me.

Anyway, for this case, using [while](http://cljs.github.io/api/cljs.core/while) might actually be a better choice, but would require more refactoring.

More importantly, however, we really shouldn't have this `hover-image` state and its manipulation in view code. The image should be a value in app-db, and set-hover should be a subscription. The subscription would simply read the existing hover-image value (from the db argument), but _not_ return it (unless only one is available)! Instead it would fetch a random image to return if different. And recur (or while) is still of use, since subscriptions are just functions, after all.